File: cmake/redirect.rb

require 'open3'

@in_file = nil
@out_file = nil
@err_file = nil
@command = nil

debug = false

args = ARGV.dup
until args.empty?
  arg = args.first
  case arg
  when "--in"
    args.shift
    @in_file = args.shift
  when "--out"
    args.shift
    @out_file = args.shift
  when "--err"
    args.shift
    @err_file = args.shift
  else
    @command = args.dup
    @command[0] = File.absolute_path(@command[0])
    args = []
  end
end

raise "Missing COMMAND argument" if @command.nil?

if @in_file
  stdin_data = File.read(@in_file)
else
  stdin_data = nil
end

out,err,status = Open3.capture3(*@command, :stdin_data => stdin_data)

if @out_file
  File.open(@out_file, "w") {|o| o.write out}
end
if @err_file
  File.open(@err_file, "w") {|o| o.write err}
end

if !status.success?
  if debug
    STDERR.puts File.read(@err_file)
  end
  exit -1
end

File: cmake/redirect.py
# Souffle - A Datalog Compiler
# Copyright (c) 2022 The Souffle Developers. All rights reserved
# Licensed under the Universal Permissive License v 1.0 as shown at:
# - https://opensource.org/licenses/UPL
# - <souffle root>/licenses/SOUFFLE-UPL.txt

import os
import argparse
import pathlib
import subprocess

parser = argparse.ArgumentParser(description="Redirect standard streams")
parser.add_argument('--in', dest='in_file')
parser.add_argument('--out', dest='out_file')
parser.add_argument('--err', dest='err_file')
parser.add_argument('command', type=lambda p: pathlib.Path(p).absolute())
parser.add_argument('arguments', nargs=argparse.REMAINDER)

args = parser.parse_args()

if args.in_file:
    stdin = open(args.in_file)
else:
    stdin = None

if args.out_file:
    stdout = open(args.out_file, "w")
else:
    stdout = None

if args.err_file:
    stderr = open(args.err_file, "w")
else:
    stderr = None

status = subprocess.run([args.command] + args.arguments, stdin=stdin, stdout=stdout, stderr=stderr)

if stdout:
    stdout.close()

if stderr:
    stderr.close()

os.sys.exit(status.returncode)
